providers/common/macs/gmac_prov.c

```diff
@@ -0,0 +1,289 @@
+    OSSL_PARAM_octet_string(OSSL_MAC_PARAM_KEY, NULL, 0),
+    OSSL_PARAM_octet_string(OSSL_MAC_PARAM_IV, NULL, 0),
+    OSSL_PARAM_END
+};
+static const OSSL_PARAM *gmac_ctx_set_param_types(void)
+{
+    return known_ctx_set_params;
+}
+
+/*
+ * ALL parameters should be set before init().
+ */
+static int gmac_ctx_set_params(void *vmacctx, const OSSL_PARAM params[])
+{
+    struct gmac_data_st *macctx = vmacctx;
+    EVP_CIPHER_CTX *ctx = macctx->ctx;
+    const OSSL_PARAM *p;
+
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_MAC_PARAM_CIPHER)) != NULL
+        || (p = OSSL_PARAM_locate_const(params,
+                                        OSSL_MAC_PARAM_ALGORITHM)) != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return 0;
+
+        {
+            const char *algoname = p->data;
+            const char *propquery = NULL;
+
+#ifndef FIPS_MODE /* Inside the FIPS module, we don't support engines */
+            ENGINE_finish(macctx->engine);
+            macctx->engine = NULL;
+
+            if ((p = OSSL_PARAM_locate_const(params, OSSL_MAC_PARAM_ENGINE))
+                != NULL) {
+                if (p->data_type != OSSL_PARAM_UTF8_STRING)
+                    return 0;
+
+                macctx->engine = ENGINE_by_id(p->data);
+                if (macctx->engine == NULL)
+                    return 0;
+            }
+#endif
+            if ((p = OSSL_PARAM_locate_const(params,
+                                             OSSL_MAC_PARAM_PROPERTIES))
+                != NULL) {
+                if (p->data_type != OSSL_PARAM_UTF8_STRING)
+                    return 0;
+
+                propquery = p->data;
+            }
+
+            EVP_CIPHER_meth_free(macctx->alloc_cipher);
+            macctx->cipher = macctx->alloc_cipher = NULL;
+
+            macctx->cipher = macctx->alloc_cipher =
+                EVP_CIPHER_fetch(PROV_LIBRARY_CONTEXT_OF(macctx->provctx),
+                                 algoname, propquery);
+#ifndef FIPS_MODE /* Inside the FIPS module, we don't support legacy ciphers */
+            /* TODO(3.0) BEGIN legacy stuff, to be removed */
+            if (macctx->cipher == NULL)
+                macctx->cipher = EVP_get_cipherbyname(algoname);
+            /* TODO(3.0) END of legacy stuff */
+#endif
+
+            if (macctx->cipher == NULL)
+                return 0;
+
+            if (EVP_CIPHER_mode(macctx->cipher) != EVP_CIPH_GCM_MODE) {
+                ERR_raise(ERR_LIB_PROV, EVP_R_CIPHER_NOT_GCM_MODE);
+                return 0;
+            }
+        }
+        if (!EVP_EncryptInit_ex(ctx, macctx->cipher, macctx->engine,
+                                NULL, NULL))
+            return 0;
+    }
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_MAC_PARAM_KEY)) != NULL) {
+        if (p->data_type != OSSL_PARAM_OCTET_STRING)
+            return 0;
+
+        if (p->data_size != (size_t)EVP_CIPHER_CTX_key_length(ctx)) {
```

Comment:
Are we exposing this function to the FIPS module?

Reply:
Yup.  It's implemented in `crypto/evp/evp_lib.c`, which is used in the FIPS module.

Also, we build all modules with `-z defs` on Linux (and others too, but Linux is what I remember), so if a function isn't exposed to the FIPS module, the FIPS module build will fail.